src/Stream.sol

```diff
@@ -228,110 +238,93 @@ contract Stream is IStream, Clone {
+     * @dev Checking token balance before and after to defend against the case of multiple contracts
+     * updating the balance of the same token.
      * @param tokenAddress the contract address of the token to recover.
      * @param amount the amount to recover.
      */
-    function rescueERC20(address tokenAddress, uint256 amount) external onlyPayer {
-        if (tokenAddress == address(token())) revert CannotRescueStreamToken();
+    function recoverTokens(address tokenAddress, uint256 amount) external onlyPayer {
+        // When the stream is under-funded, it should keep its current balance
+        // When it's sufficiently-funded, it should keep the full balance committed to recipient
+        // i.e. `remainingBalance` or `recipientCancelBalance`
+        uint256 requiredBalanceAfter =
+            Math.min(tokenBalance(), Math.max(remainingBalance, recipientCancelBalance));
 
         IERC20(tokenAddress).safeTransfer(msg.sender, amount);
-    }
 
-    /**
-     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
-     *   VIEW FUNCTIONS
-     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
-     */
+        if (tokenBalance() < requiredBalanceAfter) revert RescueTokenAmountExceedsExcessBalance();
 
-    /**
-     * @notice Returns the available funds to withdraw.
-     * @param who The address for which to query the balance.
-     * @return uint256 The total funds allocated to `who` as uint256.
-     */
-    function balanceOf(address who) public view returns (uint256) {
-        uint256 recipientBalance = _recipientBalance();
-
-        if (who == recipient()) return recipientBalance;
-        if (who == payer()) {
-            // This is safe because it should always be the case that:
-            // remainingBalance >= recipientBalance.
-            unchecked {
-                return remainingBalance - recipientBalance;
-            }
-        }
-        return 0;
-    }
-
-    /**
-     * @notice Returns the time elapsed in this stream, or zero if it hasn't started yet.
-     */
-    function elapsedTime() public view returns (uint256) {
-        uint256 startTime_ = startTime();
-        if (block.timestamp <= startTime_) return 0;
-
-        uint256 stopTime_ = stopTime();
-        if (block.timestamp < stopTime_) return block.timestamp - startTime_;
-
-        return stopTime_ - startTime_;
-    }
-
-    /**
-     * @notice Get this stream's token balance vs the token amount required to meet the commitment
-     * to recipient.
-     */
-    function tokenAndOutstandingBalance() public view returns (uint256, uint256) {
-        return (tokenBalance(), remainingBalance);
+        emit TokensRecovered(tokenAddress, amount);
     }
 
     /**
      * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
-     *   INTERNAL FUNCTIONS
+     *   VIEW FUNCTIONS
      * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
      */
 
     /**
      * @dev Helper function for `balanceOf` in calculating recipient's fair share of tokens, taking withdrawals into account.
```

Comment:
update the natspec, it's no longer a helper function

Reply:
fixed